File: mediacontentprovider/src/main/java/com/mrlukashem/mediacontentprovider/content/MediaContentViewBuilder.java
package com.mrlukashem.mediacontentprovider.content;

import com.mrlukashem.mediacontentprovider.types.ContentType;
import com.mrlukashem.mediacontentprovider.types.MediaContentField;

import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;

public class MediaContentViewBuilder {
  private Map<MediaContentField, String> initialFields = new HashMap<>();
  private ContentType contentTypeToBuild;

  public MediaContentViewBuilder(@NotNull ContentType contentTypeToBuild) {
    this.contentTypeToBuild = contentTypeToBuild;
  }

  public IMediaContentView build() {
    return new MediaContentView(initialFields, contentTypeToBuild);
  }

  public MediaContentViewBuilder addField(@NotNull MediaContentField name, @NotNull String value) {
    initialFields.put(name, value);
    return this;
  }
}

File: mediacontentprovider/src/main/java/com/mrlukashem/mediacontentprovider/content/MediaContentViewBuilder.kt
package com.mrlukashem.mediacontentprovider.content

import com.mrlukashem.mediacontentprovider.types.ContentType
import com.mrlukashem.mediacontentprovider.types.MediaContentField

import java.util.HashMap

class MediaContentViewBuilder(private val contentTypeToBuild: ContentType) {
    var initialFields = ArrayList<MediaContentField>()

    constructor(mainType : ContentType.MainType, subType: ContentType.SubType)
            : this(ContentType(mainType, subType)) {}

    fun build(): IMediaContentView {
        return MediaContentView(initialFields, contentTypeToBuild)
    }

    fun addField(field : MediaContentField) : MediaContentViewBuilder {
        initialFields.add(field)
        return this
    }

    fun addField(fieldName: MediaContentField.FieldName, fieldValue: String)
            : MediaContentViewBuilder {
        return addField(MediaContentField(fieldName, fieldValue))
    }
}
